graylog2-server/src/main/java/org/graylog2/indexer/datanode/DatanodeMigrationLockServiceImpl.java

```diff
@@ -0,0 +1,127 @@
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+public class DatanodeMigrationLockServiceImpl implements DatanodeMigrationLockService {
+
+    private static final Logger LOG = LoggerFactory.getLogger(DatanodeMigrationLockServiceImpl.class);
+    public static final int LOCK_EXTEND_PERIOD_SECONDS = 10;
+    private static final String LOCK_RESOURCE_PREFIX = "remote-reindex-migration_";
+
+    private final LockService lockService;
+
+
+      private final Set<Lock> activeLocks = Collections.synchronizedSet(new HashSet<>());
+
+    @Inject
+    public DatanodeMigrationLockServiceImpl(LockService lockService) {
+        this.lockService = lockService;
+        startLocksExtendingThread(lockService);
+    }
+
+    private void startLocksExtendingThread(LockService lockService) {
+        ScheduledExecutorService executorService = Executors.newSingleThreadScheduledExecutor(new ThreadFactoryBuilder().setNameFormat("migration-locks-service-backend-%d").setDaemon(true).setUncaughtExceptionHandler(new Tools.LogUncaughtExceptionHandler(LOG)).build());
+        executorService.scheduleAtFixedRate(() -> {
+            final Set<Optional<Lock>> extendedLocks = activeLocks.stream().map(lockService::extendLock).collect(Collectors.toSet());
+            if (!extendedLocks.isEmpty()) {
+                LOG.info("Extended TTL of {} datanode migration locks", extendedLocks.size());
+            }
+        }, LOCK_EXTEND_PERIOD_SECONDS, LOCK_EXTEND_PERIOD_SECONDS, TimeUnit.SECONDS);
+    }
+
+    @Override
+    public Lock acquireLock(IndexSet indexSet, Class<?> caller, DatanodeMigrationLockWaitConfig config) {
+        final String indexSetID = indexSet.getConfig().id();
+        final String resource = LOCK_RESOURCE_PREFIX + indexSetID;
+        return waitForLock(resource, caller, indexSet, config);
+    }
+
+    @Override
+    public void tryRun(IndexSet indexSet, Class<?> caller, Runnable runnable) {
+        final Optional<Lock> lock = tryLock(indexSet, caller);
+        // here we have to keep the lock refreshed every now and then
+        lock.ifPresentOrElse(l -> {
+            try {
+                runnable.run();
+            } finally {
+                release(l);
+            }
+        }, () -> LOG.info("Couldn't enquire a lock of index set {}({}) for {}, skipping execution", indexSet.getConfig().title(), indexSet.getConfig().id(), caller.getName()));
+    }
+
+    @Override
+    public void release(Lock lock) {
+        activeLocks.remove(lock);
+        lockService.unlock(lock);
+    }
+
+    private Optional<Lock> tryLock(IndexSet indexSet, Class<?> caller) {
+        final String resource = LOCK_RESOURCE_PREFIX + indexSet.getConfig().id();
+        return doLock(resource, caller);
+    }
+
+    private synchronized Optional<Lock> doLock(String resource, Class<?> caller) {
+        final Optional<Lock> lock = lockService.lock(resource, caller.getName());
```

Comment:
Using the caller name as lock context makes the lock reentrant so the same caller can lock the same resource multiple times. Just asking if that is intentional. (because the lock API is confusing and I am struggling every time I use it :smile:)

Reply:
Good question, thanks! I think for this usage it doesn't matter much, as I am only trying to prevent different callers to access the same index set (migration vs rotation&retention tasks). But it could be nice to make the lock completely exclusive, I guess, maybe to prevent two migrations of the same index set in the same time. I'll extend the context with the migration ID. For the other background tasks I'd rely on mechanisms that kept it running collision free so far. Does it sound reasonable?

Reply:
Implemented in https://github.com/Graylog2/graylog2-server/pull/20222/commits/bcfe55cca7660905009b6ad3a3f759395cb2f770

Reply:
Sounds good. :+1: